doc/barclamp.md

```diff
@@ -0,0 +1,1177 @@
+# Barclamp development guide
+
+This document is an introduction to Barclamp development. It will guide you
+through all the steps neccessary to create a minimal (without UI elements)
+Barclamp for Crowbar. It will then guide you through all the testing, CI and
+review steps until the point where it gets merged into Crowbar.
+
+## Terms and definitions
+
+We will assume throughout this text that your barclamp is will be named
+`mybarclamp`. Please substitute your actual barclamp's name for this
+placeholder when you create a barclamp. The convention for OpenStack barclamps
+is to use the OpenStack component's name, e.g. `nova`.
+
+## Prerequisites
+
+We will assume some things to already be present. Namely, the following:
+
+* A Github fork of the appropriate Barclamp collection. For instance, you will
+  find OpenStack related barclamps in the
+  [crowbar-openstack](https://www.github.com/crowbar/crowbar-openstack)
+  repository and high availability related Barclamps in the
+  [crowbar-ha](https://github.com/crowbar/crowbar-ha) repository. For the sake
+  of readability we will assume you are developing an OpenStack related and
+  forked [crowbar-openstack](https://www.github.com/crowbar/crowbar-openstack)
+  throughout the rest of this document.
+
+* A Crowbar admin node with the Crowbar version you are developing against
+  (for new barclamps this will be `master`) already installed and running. This
+  should be as up-to-date as possible to reduce the chance of merge conflicts
+  happening.
+
+* A set of machines for your Crowbar admin node to orchestrate. In the case of
+  SUSE OpenStack Cloud this would be a controller and one compute node at a
+  minimum.
+
+Providing these is out of scope for this document. We assume you are able to
+fork the appropriate Barclamp collection. The machines can be provided through
+manual installation or by using an automated setup tool such as
+[mkcloud](https://github.com/SUSE-Cloud/automation/blob/master/scripts/mkcloud).
+
+Optionally (but strongly recommended) you can also
+
+* Create a backup of your Crowbar admin node: You are probably going to break
+  it a few times in the course of testing your new barclamp so it is a good
+  idea to have a backup you can quickly restore rather than rebuild your entire
+  test environment. If you are using `mkcloud` you can create and restore a LVM
+  snapshot of your admin node using its `createadminsnapshot` and
+  `restoreadminfromsnapshot` steps, respectively.
+
+* Create an additional OpenStack controller or compute node as a scratch node:
```

Comment:
Looks good. Thanks :-)

I added this to the documentation in slightly modified form.